File: solutions/javascript/phone-number/1/phone-number.js
//
// This is only a SKELETON file for the 'Phone Number' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const clean = (number) => {
  let cleaned_number = "";
  for(let char of number) {
    if(/^[0-9]$/.test(char))
      cleaned_number += char
    else if(/^[a-zA-Z]$/.test(char))
      throw new Error('Letters not permitted');
    else if(!"()-+ .".includes(char))
      throw new Error('Punctuations not permitted');
  }
  const l = cleaned_number.length;
  if(l < 10)
    throw new Error('Incorrect number of digits');
  if(l > 11)
    throw new Error('More than 11 digits');
  if(l === 11){
    if(cleaned_number[0] !== "1")
      throw new Error('11 digits must start with 1');
    cleaned_number = cleaned_number.slice(1);
  }
  if(cleaned_number[0] === "1")
    throw new Error('Area code cannot start with one');
  if(cleaned_number[0] === "0")
    throw new Error('Area code cannot start with zero');
  if(cleaned_number[3] === "1")
    throw new Error('Exchange code cannot start with one');
  if(cleaned_number[3] === "0")
    throw new Error('Exchange code cannot start with zero');
  return cleaned_number;
};

File: solutions/python/phone-number/1/phone_number.py
class PhoneNumber:
    def __init__(self, number):
        this_number = ""
        for char in number:
            if char.isdigit():
                this_number += char
            elif char.isalpha():
                raise ValueError("letters not permitted")
            elif not(char in "()-. +"):
                raise ValueError("punctuations not permitted")
        l = len(this_number)
        if l < 10:
            raise ValueError("must not be fewer than 10 digits")
        if l > 11:
            raise ValueError("must not be greater than 11 digits")
        if l == 11:
            if this_number[0] != "1":
                raise ValueError("11 digits must start with 1")
            this_number = this_number[1:]
        if this_number[3] == "1":
            raise ValueError("exchange code cannot start with one")
        if this_number[3] == "0":
            raise ValueError("exchange code cannot start with zero")
        if this_number[0] == "1":
            raise ValueError("area code cannot start with one")
        if this_number[0] == "0":
            raise ValueError("area code cannot start with zero")
        self.number = this_number
        self.area_code = this_number[:3]

    def pretty(self):
        return f"({self.area_code})-{self.number[3:6]}-{self.number[6:]}"


